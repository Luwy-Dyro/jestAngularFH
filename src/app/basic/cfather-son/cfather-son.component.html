<h5>Component Soon</h5>


<code>{{client | json}}</code>

<br><br>

<button
    class="btn btn-outline-primary"
    (click)="onChange(10)"
    data-test="btn-changeId"

>
Cambiar ID

</button>

<button
    data-test="btn-delete"
    class="btn btn-outline-danger"
    (click)="onDelete()"
>
Borrar Client

</button>