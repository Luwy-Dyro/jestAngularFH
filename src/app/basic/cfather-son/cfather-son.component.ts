import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Client } from '../interfaces/client';

@Component({
  selector: 'app-cfather-son',
  templateUrl: './cfather-son.component.html',
  styleUrls: ['./cfather-son.component.scss']
})
export class CfatherSonComponent {

  @Input() client?: Client
  @Output() onDeleteClient = new EventEmitter();
  @Output() onUpdatedClient = new EventEmitter<Client>() ;

  onDelete(){
    this.client = undefined;
    this.onDeleteClient.emit()
  }

  onChange(newId: number){

    if(!this.client) return;

    //this.client.id = newId
    this.client = {
      ...this.client,
      id: newId
    }

    this.onUpdatedClient.emit({...this.client})


  }

}
