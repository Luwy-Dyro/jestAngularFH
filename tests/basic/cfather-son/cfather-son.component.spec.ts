import { ComponentFixture, TestBed } from '@angular/core/testing';

import { CfatherSonComponent } from '../../../src/app/basic/cfather-son/cfather-son.component';

describe('CfatherSonComponent', () => {
  let component: CfatherSonComponent;
  let fixture: ComponentFixture<CfatherSonComponent>;

  let compiled: HTMLElement

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ CfatherSonComponent ]
    })
    .compileComponents();

    fixture = TestBed.createComponent(CfatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    compiled = fixture.nativeElement
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });


  it('debe hacer match con snapshot', () => {
   
    expect(compiled).toMatchSnapshot();

  });

  it('no debe mostrar btn si no hay client', () => {
    
    const button = compiled.querySelectorAll('button')
    expect(button.length).toBe(2);

  });


  it('debe mostrar 2 btn si hay client', () => {
    component.client = { id: 1, name: 'Luis' }
    fixture.detectChanges()

    // expect(compiled).toMatchSnapshot();
    const button = compiled.querySelectorAll('button')
    expect(button.length).toBe(2);

  });

  it('debe emitir onDeleteClient con el btn elim', () => {
    component.client = { id: 1, name: 'Luis' }
    fixture.detectChanges()

    //Pasa a ser una funcion de JEST - saber con que argumento se llamo, las veces q se llamo...
    jest.spyOn( component.onDeleteClient, 'emit' );

    const btnDelete = compiled.querySelector('[data-test=btn-delete]')
    //console.log(btnDelete?.textContent );

    btnDelete?.dispatchEvent(new Event('click'));

    // expect(component.onDeleteClient.emit).toHaveBeenCalledWith('hola mundo')
    expect(component.onDeleteClient.emit).toHaveBeenCalled()
    // expect(button.length).toBe(0);

  });


  it('debe emitir onUpdatedClient con el btn Cambiar ID', () => {
    component.client = { id: 1, name: 'Luis' }
    fixture.detectChanges()
    
    jest.spyOn( component.onUpdatedClient, 'emit' );

    const btnChangeId = compiled.querySelector('[data-test=btn-changeId]')

    //Solo esta probando el click con todo sus valores
    btnChangeId?.dispatchEvent(new Event('click'));

    expect(component.onUpdatedClient.emit).toHaveBeenCalledWith({
      id: 10,
      name: 'Luis'
    })
    // expect(button.length).toBe(0);

  });



  it('debe emitir onChangeClient con el ID especificado SI hay un cliente', () => {
    
    
    jest.spyOn( component.onUpdatedClient, 'emit' )

    component.onChange(10)

    expect(component.onUpdatedClient.emit).not.toHaveBeenCalled()

    component.client = { id: 1, name: 'Luis' }
    fixture.detectChanges()
    component.onChange(10)

    expect(component.onUpdatedClient.emit).toHaveBeenLastCalledWith({

      id: 10,
      name: 'Luis'

    })


  });


});
