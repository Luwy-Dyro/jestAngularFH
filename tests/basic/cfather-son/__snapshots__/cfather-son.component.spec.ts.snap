// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CfatherSonComponent debe hacer match con snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Component Soon
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  <button
    class="btn btn-outline-primary"
    data-test="btn-changeId"
  >
    
Cambiar ID 
  </button>
  <button
    class="btn btn-outline-danger"
    data-test="btn-delete"
  >
    
Borrar Client 
  </button>
</div>
`;
